exporters/failure/collector_azure_devops.py

```diff
@@ -0,0 +1,180 @@
+#    License for the specific language governing permissions and limitations
+#    under the License.
+#
+
+import logging
+from typing import Iterator
+
+from attrs import converters, define, field
+from azure.devops.connection import Connection
+from azure.devops.exceptions import AzureDevOpsServiceError
+from azure.devops.v6_0.work_item_tracking.models import Wiql, WorkItem
+from msrest.authentication import BasicAuthentication
+
+from failure.collector_base import AbstractFailureCollector, TrackerIssue
+from pelorus.config import env_var_names, env_vars
+from pelorus.config.converters import comma_or_whitespace_separated, pass_through
+from pelorus.config.log import REDACT, log
+from pelorus.errors import FailureProviderAuthenticationError
+from pelorus.timeutil import parse_assuming_utc, second_precision
+from pelorus.utils import Url
+
+_DATETIME_FORMAT = "%Y-%m-%dT%H:%M:%S.%fZ"
+
+
+@define(kw_only=True)
+class AzureDevOpsFailureCollector(AbstractFailureCollector):
+    """
+    Azure DevOps implementation of a FailureCollector
+    """
+
+    token: str = field(
+        default="",
+        metadata=env_vars(*env_var_names.TOKEN) | log(REDACT),
+        repr=False,
+    )
+
+    tracker_api: Url = field(
+        metadata=env_vars("SERVER"),
+        converter=converters.optional(pass_through(Url, Url.parse)),
+    )
+
+    projects: set[str] = field(
+        factory=set, converter=comma_or_whitespace_separated(set)
+    )
+
+    work_item_type: set[str] = field(
+        factory=set,
+        converter=comma_or_whitespace_separated(set),
+        metadata=env_vars("AZURE_DEVOPS_TYPE"),
+    )
+
+    work_item_priority: set[str] = field(
+        factory=set,
+        converter=comma_or_whitespace_separated(set),
+        metadata=env_vars("AZURE_DEVOPS_PRIORITY"),
+    )
+
+    def __attrs_post_init__(self):
+        if self.token:
+            credentials = BasicAuthentication("", self.token)
+            connection = Connection(base_url=self.tracker_api.url, creds=credentials)
+            self.client = connection.clients_v6_0.get_work_item_tracking_client()
+
+    def get_work_items(self) -> Iterator[WorkItem]:
+        logging.info("Getting work items")
+
+        try:
+            wiql = Wiql(query="Select [System.Id] From WorkItems")
+            wiql_results = self.client.query_by_wiql(wiql).work_items
+            if wiql_results:
+                # WIQL query gives a WorkItemReference with ID only
+                # get the corresponding WorkItem from ID
+                work_items = (
+                    self.client.get_work_item(int(res.id)) for res in wiql_results
+                )
+            return work_items
+        except AzureDevOpsServiceError as error:
+            logging.error(FailureProviderAuthenticationError.auth_message)
+            raise FailureProviderAuthenticationError from error
+        except Exception as error:
+            # TODO test limitation 200?!?!
```

Comment:
TODO https://learn.microsoft.com/en-us/rest/api/azure/devops/wit/work-items/list?view=azure-devops-rest-7.0&tabs=HTTP

Reply:
since we are using wiql endpoint, there seems to be no limitation